storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1753 @@ buf_block_t *fsp_get_latched_xdes_page(
+
+  (2) If the xdes_get_stats(m_old_descr) is
+      XDES_FREE_FRAG / XDES_FULL_FRAG then
+
+     (2.1) If the number of used pages in extent descriptor
+           is FSP_EXTENT_SIZE then move the extent descriptor
+           from FSP_FULL_FRAG to FSP_FREE_FRAG list
+     (2.2) If the number of used pages in extent descriptor is 0 then
+           move the extent descriptor from FSP_FREE_FRAG to FSP_FREE
+     (2.3) Free the page and mark the XDES_FREE_BIT of the respective
+           page in current extent descriptor
+Above all scenario done by 2 steps to make sure that there
+will be no error scenario once the modification of the pages
+has started.
+1) prepare - Basically validates the necessary condition
+and make sure that pages are being latched
+2) Complete - Completes the action by using the latched
+pages in prepare step */
+class PageOperator final
+{
+  /** Header block for the tablespace */
+  buf_block_t *const m_header_block= nullptr;
+  /** Index node block */
+  buf_block_t *const m_iblock= nullptr;
+  /** Index node */
+  fseg_inode_t *const m_inode= nullptr;
+  /** offset of index node within index node page*/
+  uint16_t m_ioffset= 0;
+  /** Maximum free limit of the tablespace */
+  uint32_t m_free_limit= 0;
+  /** Segment id */
+  uint64_t m_seg_id= 0;
+  /** Extent size */
+  uint32_t m_extent_size= 0;
+
+  /** New block to be allocated */
+  buf_block_t *m_new_block= nullptr;
+  /** New block extent descriptor */
+  buf_block_t *m_new_xdes= nullptr;
+  /** New block descriptor */
+  xdes_t *m_new_descr= nullptr;
+  /** New block descriptor offset within xdes page */
+  uint16_t m_xoffset= 0;
+  /** New extent descriptor state */
+  uint32_t m_new_state= 0;
+  /** Need segment allocation */
+  bool m_need_segment= false;
+  /** Old pages during allocation to be saved */
+  buf_block_t *m_old_pages[8]= {nullptr};
+  /** Page to be removed */
+  byte m_old_page_no[4]= {0};
+  /** Old block extent descriptor page */
+  buf_block_t *m_old_xdes= nullptr;
+  /** Old block descriptor */
+  xdes_t *m_old_descr= nullptr;
+  /** Old block descriptor offset with descriptor page */
+  uint16_t m_old_xoffset= 0;
+  /** Old descriptor state */
+  uint32_t m_old_state= 0;
+  /** Mini-transaction to allocate & free a page */
+  mtr_t *const m_mtr;
+
+  /** Save the old page state of the block before
+  allocating a page
+  @param block   block to be stored
+  @return error code */
+  dberr_t save_old_page(buf_block_t *block) noexcept
+  {
+    if (!block) return DB_SUCCESS;
+    uint32_t first_free= UINT32_MAX;
+    for (uint32_t i= 0; i < 8; i++)
+    {
+      if (m_old_pages[i] == nullptr)
+      {
+        if (first_free == UINT32_MAX)
+          first_free= i;
+        continue;
+      }
+
+      if (memcmp(m_old_pages[i]->page.frame + FIL_PAGE_OFFSET,
+                 block->page.frame + FIL_PAGE_OFFSET, 4) == 0)
+        return DB_SUCCESS;
+    }
+
+    if (first_free == UINT32_MAX)
+      return DB_CORRUPTION;
```

Comment:
This can be simplified further:
```c++
  size_t first_free;
  for (first_free= 0; first_free < array_elements(m_old_pages); first_free++)
  {
    const buf_block_t *b= m_old_pages[first_free];
    if (!b)
      goto found;
    if (b->page.hash == &block->page)
      return DB_SUCCESS;
  }
  return DB_CORRUPTION;
found:
```